cmd/influx_inspect/buildtsi/buildtsi.go

```diff
@@ -182,40 +183,37 @@ func (cmd *Command) compactSeriesFilePartition(path string) error {
 	const tmpExt = ".tmp"
 
 	fmt.Fprintf(cmd.Stdout, "processing partition for %q\n", path)
-	fis, err := ioutil.ReadDir(path)
+
+	// Open partition so index can recover from entries not in the snapshot.
+	partitionID, err := strconv.Atoi(filepath.Base(path))
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot parse partition id from path: %s", path)
 	}
-
-	indexPath := filepath.Join(path, "index")
-	index := tsdb.NewSeriesIndex(indexPath)
-	if err := index.Open(); err != nil {
-		return err
+	p := tsdb.NewSeriesPartition(partitionID, path, nil)
+	if err := p.Open(); err != nil {
+		return fmt.Errorf("cannot open partition: path=%s err=%s", path, err)
 	}
-	defer index.Close()
-
-	for _, fi := range fis {
-		segmentID, err := tsdb.ParseSeriesSegmentFilename(fi.Name())
-		if err != nil {
-			continue // skip non-segment file.
-		}
+	defer p.Close()
 
-		segmentPath := filepath.Join(path, fi.Name())
-		fmt.Fprintf(cmd.Stdout, "processing segment %q %d\n", path, segmentID)
+	// Loop over segments and compact.
+	indexPath := p.IndexPath()
+	var segmentPaths []string
+	for _, segment := range p.Segments() {
+		fmt.Fprintf(cmd.Stdout, "processing segment %q %d\n", segment.Path(), segment.ID())
 
-		segment := tsdb.NewSeriesSegment(segmentID, path)
-		if err = segment.Open(); err != nil {
-			return err
-		} else if err := segment.CompactToPath(segmentPath+tmpExt, index); err != nil {
-			return err
-		} else if err := segment.Close(); err != nil {
+		if err := segment.CompactToPath(segment.Path()+tmpExt, p.Index()); err != nil {
 			return err
 		}
+		segmentPaths = append(segmentPaths, segment.Path())
+	}
```

Comment:
Loops over segment file list from `Partition` instead of using `ReadDir()` as the directory can include index files (or `.tmp` from previous runs).